frontend/lib/resolution/Resolver.cpp

```diff
@@ -3746,10 +3809,55 @@ bool Resolver::enter(const uast::Domain* decl) {
 }
 
 void Resolver::exit(const uast::Domain* decl) {
+  if (scopeResolveOnly) {
+    return;
+  }
+
+  const DomainType* genericDomainType = DomainType::getGenericDomainType(context);
+  if (CompositeType::isMissingBundledRecordType(context, genericDomainType->id())) {
+    // If we don't have the standard library code backing the Domain type, leave
+    // it unresolved.
+    return;
+  }
+
   if (decl->numExprs() == 0) {
+    // Generic domain, as in a generic-domain array
     auto& re = byPostorder.byAst(decl);
-    auto dt = QualifiedType(QualifiedType::CONST_VAR, DomainType::getGenericDomainType(context));
+    auto dt = QualifiedType(QualifiedType::CONST_VAR, genericDomainType);
     re.setType(dt);
+  } else {
+    // Call appropriate domain builder proc. Use ensureDomainExpr when the
+    // domain is declared without curly braces (within an array type).
+    const char* domainBuilderProc = decl->usedCurlyBraces()
+                                        ? "chpl__buildDomainExpr"
+                                        : "chpl__ensureDomainExpr";
+
+    // Add key or range actuals
+    std::vector<CallInfoActual> actuals;
+    for (auto expr : decl->exprs()) {
+      actuals.emplace_back(byPostorder.byAst(expr).type(), UniqueString());
+    }
+
+    // Add definedConst actual if appropriate
+    if (decl->usedCurlyBraces()) {
+      actuals.emplace_back(
+          QualifiedType(QualifiedType::PARAM, BoolType::get(context),
+                        BoolParam::get(context, true)),
+          UniqueString());
+    }
```

Comment:
Not sure actually, I did it because it's [what we do in `convert-uast`](https://github.com/chapel-lang/chapel/blob/939475fa0defad3f741d43f18b3913afbb05abbb/compiler/passes/convert-uast.cpp#L2088-L2090). I'll look for a better justification than that

Reply:
I figured it was a historical reason, and I don't think the onus should be on you to figure out why this is the case, but it would definitely satisfy my curiosity.

Reply:
Here's my tiny investigation:

- We first started setting domain literals `definedConst` always true in https://github.com/chapel-lang/chapel/pull/16397, and carried it over to `convert-uast` in https://github.com/chapel-lang/chapel/pull/17975.
- Looks like we adjust `definedConst` via [later modifications to AST as we hit usages of the domain](https://github.com/chapel-lang/chapel/blob/main/compiler/optimizations/propagateDomainConstness.cpp#L263), and from the OP of that first PR I gather it was simpler to impleent starting off as true.
- I didn't look more into details of the later adjustments, but I think there's a good chance the production compiler is still relying on domain literals starting off with `definedConst` true.